crates/core/tedge_api/src/device_id.rs

```diff
@@ -0,0 +1,36 @@
+pub fn new_get_child_id_from_topic(parent_device_name: String, topic: String) -> Option<String> {
+    if topic.starts_with("te/device") {
+        let topic_split = topic.split('/').collect::<Vec<_>>();
+        if topic_split.len() >= 5 {
+            if topic_split[2].eq("main") {
+                return None;
+            }
```

Comment:
There seems to be a logic error here.

With the above logic, the following two topics will return the same thing, even though they are very different (one is for the main device, and the other for a service on the main device).

```
te/device/main// => None
te/device/main/service/nodered => None
```

Also I would avoid using `None` as a reference to the device certificate's common name...it would just be better to return the `parent_device_name` variable instead (imo).

Reply:
I would have expected a function more like this (though I'm not a Rust dev, so forgot any bad-pratices):

```rust
fn get_external_identity(topic: String, root_prefix: String, common_name: String) -> Option<String> {
    let sep = ":";
    
    let local_name = topic.split("/").take(5).filter(|&x| !x.is_empty()).collect::<Vec<&str>>();
    if local_name.len() == 0 {
        // ignore if all empty, as something is obviously wrong
        return None;
    }

    let main_topic = format!("{}/device/main", root_prefix);
    if local_name.join("/") == main_topic {
        return Some(common_name.to_string());
    }

    if local_name[0] == root_prefix {
        match local_name.len() {
            1 => {
                return Some(common_name.to_string());
            }
            _ => {
                let mut output = format!("{}{}", common_name, sep).to_string();
                if local_name.len() > 1 {
                    output.push_str(local_name.into_iter().skip(1).collect::<Vec<&str>>().join(sep).as_str());
                }
                return Some(output);
            }
            
        }
    } 
    return None;    
}
```

Reply:
I think the logic is right for the `telemetry` topics. When it comes to the `service` it might be wrong. That is a separate issue, not sure to address it here or later. IMO it's better to look into it later.